File: prob_35.java
public class prob_35 {
    public static void main(String args[]) {
        int nums[] = {1,3,5,6}, target = 7; 
        System.out.println(Solution_35.searchInsert(nums, target));
    }
}

class Solution_35 {
    public static int searchInsert(int[] nums, int target) {
        return searchInsertBin(nums, 0, nums.length-1, target);
    }

    public static int searchInsertBin(int[] nums, int i, int j, int target) {
        if(i == j) {
            if(nums[i] >= target) return i;
            else if(nums[i] < target) return i+1;
        }
        int mid = (i+j)/2;
        if(nums[mid] == target) return mid;
        else if(nums[mid] < target) return searchInsertBin(nums, mid+1, j, target);
        else {
            if(mid == i) return mid;
            else return searchInsertBin(nums, i, mid-1, target);
        }
    }
}

File: cpp/prob_35.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
    public:
    Solution() {
        cout << "Constructor of solution for prob_35 called" << endl;
    }
    ~Solution() {
        cout << "Destructor of solution for prob_35 called" << endl;
    }
    int searchInsert(vector<int>& nums, int target) {
        int left, right, mid;
        left = 0;
        right = nums.size() - 1;
        while (left <= right) {
            mid = (left + right) / 2;
            if (nums.at(mid) == target) return mid;
            if (nums.at(mid) < target) left = mid + 1;
            else right = mid - 1;
        }
        if (nums.at(left) < target) return left + 1;
        else return left;
    }
} solution;

int main(int argc, char* argv[]) {
    vector<int> nums{1, 3, 4, 7};
    int target = 6;
    cout << solution.searchInsert(nums, target) << endl;
    return 0;
}